project1/wire_protocol/proto/ftp_messages.proto

```diff
@@ -63,6 +63,18 @@ message PwdResponse {
 /// Request to end the session.
 message QuitRequest {}
 
+/// Request to terminate a prior command.
+message TerminateRequest {
+  /// The command ID of the command to be terminated.
+  uint32 command_id = 1;
+}
+
+/// Response to GET and PUT with the terminate ID.
+message TerminateResponse {
```

Comment:
I'm not really a fan of this naming. Based on the way the other messages are named, `TerminateResponse` implies that it's a response for a specific `TerminateRequest`.

Also, I'm not sure that even having this as a separate message is the correct way to do it. For GET actions, that would break the one-message-per-response paradigm that we've been using so far, because you'd have a `GetReponse` from the server and then a `TerminateResponse` immediately afterward.

How about instead of having this as a separate message, you add `command_id` parameter to the `GetResponse` message? (You would also have to create a `PutResponse` message with this single parameter.)

Reply:
> I'm not really a fan of this naming. Based on the way the other messages are named, `TerminateResponse` implies that it's a response for a specific `TerminateRequest`.

I am also very much not a fan of the naming, but wasn't sure of what else it could be called. Do you have any suggestions?

> Also, I'm not sure that even having this as a separate message is the correct way to do it. For GET actions, that would break the one-message-per-response paradigm that we've been using so far, because you'd have a `GetReponse` from the server and then a `TerminateResponse` immediately afterward.

With our idea, it would send the `TerminateResponse` first, before sending any other response. This way, the client can prepare itself for a forked `GetResponse` to handle as a task on the thread pool.

> How about instead of having this as a separate message, you add `command_id` parameter to the `GetResponse` message? (You would also have to create a `PutResponse` message with this single parameter.)

We initially planned on doing it like that, but we changed our minds based on our understanding of how protobuf works. We could be wrong, but we're under the impression that none of the response could be interpreted until all of the response is received by the client. In the case of a `GetResponse`, file contents would be received at the same time as the terminate ID, voiding the entire purpose of that terminate ID. However, we may be wrong as we haven't explicitly tested that.

Reply:
Yeah, actually, that didn't occur to me. I think you're right that we need two separate responses if we want to preempt the file data with the command ID.

However, I would be more in favor of doing it the reverse way: Create a separate `FileContents` message for holding the contents of large files, and still send the command ID in `GetResponse` and `PutResponse` messages.

Side note: The "correct" way to deal with large files is to break them up into chunks and send each individually. If you don't do it that way, at some point you have to start considering RAM usage on the client and server side. It really depends on exactly how big the files that this is going to be tested with are. Might be a good question for the professor...

Reply:
> However, I would be more in favor of doing it the reverse way: Create a separate `FileContents` message for holding the contents of large files, and still send the command ID in `GetResponse` and `PutResponse` messages. That entirely side-steps the naming issue, and I think it's more intuitive.

I actually really like this idea, though the `FileContents` message may have to include the command ID also to resolve potential concurrency issues.

Reply:
Yeah, I'd go ahead and assume that it doesn't for now and add that later if necessary.